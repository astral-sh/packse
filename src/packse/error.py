import textwrap
from pathlib import Path


class PackseError(Exception):
    """Base type for all Packse exceptions"""


class UserError(PackseError):
    """It's the user's fault :)"""


class RequiresExtra(UserError):
    def __init__(self, feature: str, extra: str) -> None:
        message = f"Using {feature} requires installation with extra {extra!r}"
        super().__init__(message)


class DestinationAlreadyExists(UserError):
    def __init__(self, destination: Path) -> None:
        message = f"Destination directory '{destination}' already exists"
        super().__init__(message)


class InvalidScenario(UserError):
    def __init__(self, path: Path, reason: str) -> None:
        message = f"File at '{path}' is not a valid scenario: {reason}"
        super().__init__(message)


class InvalidPackageVersion(UserError):
    """Version is not PEP compliant"""

    def __init__(self, version: str) -> None:
        message = f"Version {version!r} is not valid."
        super().__init__(message)


class FileNotFound(UserError):
    def __init__(self, path: Path) -> None:
        message = f"File '{path}' not found"
        super().__init__(message)


class BuildError(PackseError):
    def __init__(self, message, output) -> None:
        message = message + ":\n" + textwrap.indent(output, " " * 4)
        super().__init__(message)


class PublishError(PackseError):
    pass


class PublishToolError(PublishError):
    def __init__(self, message, output) -> None:
        message = message + ":\n" + textwrap.indent(output, " " * 4)
        super().__init__(message)


class PublishNoCredentials(PublishError):
    def __init__(self) -> None:
        message = (
            "No credentials found for publish!\n"
            "Provide an API token via `PACKSE_PUBLISH_PASSWORD` or disable authentication with `--anonymous`"
        )
        super().__init__(message)


class PublishAlreadyExists(PublishError):
    def __init__(self, package: str) -> None:
        message = f"Publish for '{package}' already exists"
        super().__init__(message)


class PublishRateLimit(PublishError):
    def __init__(self, package: str) -> None:
        message = f"Publish of '{package}' failed due to rate limits"
        super().__init__(message)


class PublishConnectionError(PublishError):
    def __init__(self, package: str) -> None:
        message = f"Publish of '{package}' failed due to a connection error"
        super().__init__(message)


class InvalidPublishTarget(UserError):
    def __init__(self, path: Path, reason: str) -> None:
        message = f"Publish target at '{path}' is not a valid: {reason}"
        super().__init__(message)


class ServeError(PackseError):
    def __init__(self, message: str) -> None:
        super().__init__(message)


class ServeThreadError(ServeError):
    def __init__(self, message: str) -> None:
        super().__init__(message)


class ServeCommandError(ServeError):
    def __init__(self, message: str, stderr: str) -> None:
        if stderr.strip():
            message += f":\n\n{stderr}"
        super().__init__(message)


class ServeAddressInUse(ServeError):
    def __init__(self, url: str) -> None:
        message = f"Address '{url}' already in use"
        super().__init__(message)


class ServeAlreadyRunning(ServeError):
    def __init__(self) -> None:
        message = "Server is already running"
        super().__init__(message)
