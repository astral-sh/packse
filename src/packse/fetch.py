import logging
import shutil
import subprocess
import tempfile
import time
from pathlib import Path

import pkg_resources

from packse.error import DestinationAlreadyExists

logger = logging.getLogger(__name__)


def fetch(
    dest: Path | None = None,
    ref: str | None = None,
    repo_url: str = "https://github.com/astral-sh/packse",
    repo_subdir: str = "scenarios",
    force: bool = False,
):
    start_time = time.time()
    dest = dest or (Path.cwd() / "scenarios")

    if dest.exists() and not force:
        raise DestinationAlreadyExists(dest)

    if not ref:
        ref = pkg_resources.get_distribution("packse").version
        if ref == "0.0.0":
            ref = "HEAD"

    with tempfile.TemporaryDirectory() as tmpdir:
        # Perform a sparse checkout where we only grab the `scenarios` folder
        logger.info("Cloning repository %s", repo_url)
        subprocess.check_call(
            [
                "git",
                "clone",
                "-n",
                "--filter=tree:0",
                repo_url,
                "repo",
            ],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.STDOUT,
            cwd=tmpdir,
        )

        logger.info("Checking out directory '%s' at ref %s", repo_subdir, ref)
        subprocess.check_call(
            ["git", "sparse-checkout", "set", "--no-cone", repo_subdir],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.STDOUT,
            cwd=Path(tmpdir) / "repo",
        )
        subprocess.check_call(
            ["git", "fetch", "origin", ref],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.STDOUT,
            cwd=Path(tmpdir) / "repo",
        )
        subprocess.check_call(
            ["git", "checkout", ref],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.STDOUT,
            cwd=Path(tmpdir) / "repo",
        )

        src = Path(tmpdir) / "repo" / repo_subdir
        file_count = 0
        for file in sorted(src.iterdir()):
            file_count += 1
            logger.info("Found %s", file.name)

        logger.info("Copying files into '%s'", dest)
        shutil.copytree(src, dest, dirs_exist_ok=True)

    logger.info(
        "Fetched %s files in %.2fs",
        file_count,
        time.time() - start_time,
    )
