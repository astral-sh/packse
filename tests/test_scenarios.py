"""
Tests a subset of included scenarios
"""

import pytest
from packse import __development_base_path__

from .common import snapshot_command

TEST_SCENARIOS = tuple(
    sorted(
        path
        for path in (__development_base_path__ / "scenarios" / "examples").iterdir()
    )
)
TEST_SCENARIO_IDS = tuple(path.with_suffix("").name for path in TEST_SCENARIOS)


@pytest.mark.parametrize(
    "target",
    TEST_SCENARIOS,
    ids=TEST_SCENARIO_IDS,
)
@pytest.mark.usefixtures("tmpcwd")
def test_build_test_scenarios(snapshot, target):
    assert (
        snapshot_command(
            ["build", str(target)], snapshot_stderr=False, snapshot_filesystem=True
        )
        == snapshot
    )


@pytest.mark.parametrize("target", TEST_SCENARIOS, ids=TEST_SCENARIO_IDS)
@pytest.mark.usefixtures("tmpcwd")
def test_view_test_scenarios(snapshot, target):
    assert snapshot_command(["view", str(target)], snapshot_filesystem=True) == snapshot
