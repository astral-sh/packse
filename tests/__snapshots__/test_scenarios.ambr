# serializer version: 1
# name: test_build_test_scenarios[example0]
  dict({
    'exit_code': 0,
    'filesystem': dict({
      'build/example-961b4c22/example-961b4c22-0.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-961b4c22/example-961b4c22-0.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_961b4c22"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_961b4c22"]
        
        [project]
        name = "example-961b4c22"
        version = "0.0.0"
        dependencies = [
            '''example-a-961b4c22''',
        ]
        requires-python = ">=3.8"
        description = '''This is an example scenario, in which the user depends on a single package `a` which requires `b`.'''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-961b4c22/example-961b4c22-0.0.0/src/example_961b4c22/__init__.py': '''
        __version__ = "0.0.0"
  
      ''',
      'build/example-961b4c22/example-a-961b4c22-1.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-961b4c22/example-a-961b4c22-1.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_a_961b4c22"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_a_961b4c22"]
        
        [project]
        name = "example-a-961b4c22"
        version = "1.0.0"
        dependencies = [
            '''example-b-961b4c22>1.0.0''',
        ]
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-961b4c22/example-a-961b4c22-1.0.0/src/example_a_961b4c22/__init__.py': '''
        __version__ = "1.0.0"
  
      ''',
      'build/example-961b4c22/example-b-961b4c22-1.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-961b4c22/example-b-961b4c22-1.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_b_961b4c22"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_b_961b4c22"]
        
        [project]
        name = "example-b-961b4c22"
        version = "1.0.0"
        dependencies = []
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-961b4c22/example-b-961b4c22-1.0.0/src/example_b_961b4c22/__init__.py': '''
        __version__ = "1.0.0"
  
      ''',
      'build/example-961b4c22/example-b-961b4c22-2.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-961b4c22/example-b-961b4c22-2.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_b_961b4c22"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_b_961b4c22"]
        
        [project]
        name = "example-b-961b4c22"
        version = "2.0.0"
        dependencies = []
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-961b4c22/example-b-961b4c22-2.0.0/src/example_b_961b4c22/__init__.py': '''
        __version__ = "2.0.0"
  
      ''',
      'build/example-961b4c22/example-b-961b4c22-3.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-961b4c22/example-b-961b4c22-3.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_b_961b4c22"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_b_961b4c22"]
        
        [project]
        name = "example-b-961b4c22"
        version = "3.0.0"
        dependencies = [
            '''example-c-961b4c22''',
        ]
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-961b4c22/example-b-961b4c22-3.0.0/src/example_b_961b4c22/__init__.py': '''
        __version__ = "3.0.0"
  
      ''',
      'dist/example-961b4c22/example_961b4c22-0.0.0.tar.gz': 'md5:78387c54dc1123fc169bde1ad63aa11b',
      'dist/example-961b4c22/example_a_961b4c22-1.0.0-py3-none-any.whl': 'md5:93f87d75d52b4dcac32ea670f9c7c5ab',
      'dist/example-961b4c22/example_a_961b4c22-1.0.0.tar.gz': 'md5:8ed996db89e98df9b698f9e90bec5038',
      'dist/example-961b4c22/example_b_961b4c22-1.0.0-py3-none-any.whl': 'md5:96d44ceacde16b72625ffbed26c04377',
      'dist/example-961b4c22/example_b_961b4c22-1.0.0.tar.gz': 'md5:6a6fe7b8bb871d712dc7883dc0cb0d68',
      'dist/example-961b4c22/example_b_961b4c22-2.0.0-py3-none-any.whl': 'md5:fb9ab8eb91c87cf924ce14e3ed074ce0',
      'dist/example-961b4c22/example_b_961b4c22-2.0.0.tar.gz': 'md5:82b9995b2b3738d5975416105d1ab986',
      'dist/example-961b4c22/example_b_961b4c22-3.0.0-py3-none-any.whl': 'md5:3b8a4c8aa5676269e8348b747178b381',
      'dist/example-961b4c22/example_b_961b4c22-3.0.0.tar.gz': 'md5:56f1ef2dedf69f360c55297e9d9b1f20',
      'tree': '''
        test_build_test_scenarios_exam0
        ├── build
        │   └── example-961b4c22
        │       ├── example-961b4c22-0.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── example_961b4c22
        │       │           └── __init__.py
        │       ├── example-a-961b4c22-1.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── example_a_961b4c22
        │       │           └── __init__.py
        │       ├── example-b-961b4c22-1.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── example_b_961b4c22
        │       │           └── __init__.py
        │       ├── example-b-961b4c22-2.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── example_b_961b4c22
        │       │           └── __init__.py
        │       └── example-b-961b4c22-3.0.0
        │           ├── .gitignore
        │           ├── pyproject.toml
        │           └── src
        │               └── example_b_961b4c22
        │                   └── __init__.py
        └── dist
            └── example-961b4c22
                ├── example_961b4c22-0.0.0.tar.gz
                ├── example_a_961b4c22-1.0.0-py3-none-any.whl
                ├── example_a_961b4c22-1.0.0.tar.gz
                ├── example_b_961b4c22-1.0.0-py3-none-any.whl
                ├── example_b_961b4c22-1.0.0.tar.gz
                ├── example_b_961b4c22-2.0.0-py3-none-any.whl
                ├── example_b_961b4c22-2.0.0.tar.gz
                ├── example_b_961b4c22-3.0.0-py3-none-any.whl
                └── example_b_961b4c22-3.0.0.tar.gz
        
        19 directories, 24 files
  
      ''',
    }),
    'stderr': '<not included>',
    'stdout': '''
      example-961b4c22
  
    ''',
  })
# ---
# name: test_build_test_scenarios[example1]
  dict({
    'exit_code': 0,
    'filesystem': dict({
      'build/example-toml-4a4ede4d/example-toml-4a4ede4d-0.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-toml-4a4ede4d/example-toml-4a4ede4d-0.0.0/pyproject.toml': 'md5:a531b7b0544fb70b1c705bbda8a6724c',
      'build/example-toml-4a4ede4d/example-toml-4a4ede4d-0.0.0/src/example_toml_4a4ede4d/__init__.py': '''
        __version__ = "0.0.0"
  
      ''',
      'build/example-toml-4a4ede4d/example-toml-a-4a4ede4d-1.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-toml-4a4ede4d/example-toml-a-4a4ede4d-1.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_toml_a_4a4ede4d"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_toml_a_4a4ede4d"]
        
        [project]
        name = "example-toml-a-4a4ede4d"
        version = "1.0.0"
        dependencies = [
            '''example-toml-b-4a4ede4d>1.0.0''',
        ]
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-toml-4a4ede4d/example-toml-a-4a4ede4d-1.0.0/src/example_toml_a_4a4ede4d/__init__.py': '''
        __version__ = "1.0.0"
  
      ''',
      'build/example-toml-4a4ede4d/example-toml-b-4a4ede4d-1.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-toml-4a4ede4d/example-toml-b-4a4ede4d-1.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_toml_b_4a4ede4d"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_toml_b_4a4ede4d"]
        
        [project]
        name = "example-toml-b-4a4ede4d"
        version = "1.0.0"
        dependencies = []
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-toml-4a4ede4d/example-toml-b-4a4ede4d-1.0.0/src/example_toml_b_4a4ede4d/__init__.py': '''
        __version__ = "1.0.0"
  
      ''',
      'build/example-toml-4a4ede4d/example-toml-b-4a4ede4d-2.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-toml-4a4ede4d/example-toml-b-4a4ede4d-2.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_toml_b_4a4ede4d"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_toml_b_4a4ede4d"]
        
        [project]
        name = "example-toml-b-4a4ede4d"
        version = "2.0.0"
        dependencies = []
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-toml-4a4ede4d/example-toml-b-4a4ede4d-2.0.0/src/example_toml_b_4a4ede4d/__init__.py': '''
        __version__ = "2.0.0"
  
      ''',
      'build/example-toml-4a4ede4d/example-toml-b-4a4ede4d-3.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-toml-4a4ede4d/example-toml-b-4a4ede4d-3.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_toml_b_4a4ede4d"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_toml_b_4a4ede4d"]
        
        [project]
        name = "example-toml-b-4a4ede4d"
        version = "3.0.0"
        dependencies = [
            '''example-toml-c-4a4ede4d''',
        ]
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-toml-4a4ede4d/example-toml-b-4a4ede4d-3.0.0/src/example_toml_b_4a4ede4d/__init__.py': '''
        __version__ = "3.0.0"
  
      ''',
      'dist/example-toml-4a4ede4d/example_toml_4a4ede4d-0.0.0.tar.gz': 'md5:8d2873e0d003d4213d1fecef05f5202f',
      'dist/example-toml-4a4ede4d/example_toml_a_4a4ede4d-1.0.0-py3-none-any.whl': 'md5:eb4e1a9e07841a212b8d4e06764f18e2',
      'dist/example-toml-4a4ede4d/example_toml_a_4a4ede4d-1.0.0.tar.gz': 'md5:5fff429c4c7feb9d661954e6790c4a4e',
      'dist/example-toml-4a4ede4d/example_toml_b_4a4ede4d-1.0.0-py3-none-any.whl': 'md5:7e9306e7ad1291295cc1ec816148c0ae',
      'dist/example-toml-4a4ede4d/example_toml_b_4a4ede4d-1.0.0.tar.gz': 'md5:da0883a3e714b62d8661e198a861ded9',
      'dist/example-toml-4a4ede4d/example_toml_b_4a4ede4d-2.0.0-py3-none-any.whl': 'md5:99f463cfd68eb4403a6650ee2b9bfe2c',
      'dist/example-toml-4a4ede4d/example_toml_b_4a4ede4d-2.0.0.tar.gz': 'md5:1cf9f15cb12580e98c3e6ce180348629',
      'dist/example-toml-4a4ede4d/example_toml_b_4a4ede4d-3.0.0-py3-none-any.whl': 'md5:92f22666762452ac62e5f58bb2689a3d',
      'dist/example-toml-4a4ede4d/example_toml_b_4a4ede4d-3.0.0.tar.gz': 'md5:3b25c0283bd5d269ba0cf728b1ddd40d',
      'tree': '''
        test_build_test_scenarios_exam1
        ├── build
        │   └── example-toml-4a4ede4d
        │       ├── example-toml-4a4ede4d-0.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── example_toml_4a4ede4d
        │       │           └── __init__.py
        │       ├── example-toml-a-4a4ede4d-1.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── example_toml_a_4a4ede4d
        │       │           └── __init__.py
        │       ├── example-toml-b-4a4ede4d-1.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── example_toml_b_4a4ede4d
        │       │           └── __init__.py
        │       ├── example-toml-b-4a4ede4d-2.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── example_toml_b_4a4ede4d
        │       │           └── __init__.py
        │       └── example-toml-b-4a4ede4d-3.0.0
        │           ├── .gitignore
        │           ├── pyproject.toml
        │           └── src
        │               └── example_toml_b_4a4ede4d
        │                   └── __init__.py
        └── dist
            └── example-toml-4a4ede4d
                ├── example_toml_4a4ede4d-0.0.0.tar.gz
                ├── example_toml_a_4a4ede4d-1.0.0-py3-none-any.whl
                ├── example_toml_a_4a4ede4d-1.0.0.tar.gz
                ├── example_toml_b_4a4ede4d-1.0.0-py3-none-any.whl
                ├── example_toml_b_4a4ede4d-1.0.0.tar.gz
                ├── example_toml_b_4a4ede4d-2.0.0-py3-none-any.whl
                ├── example_toml_b_4a4ede4d-2.0.0.tar.gz
                ├── example_toml_b_4a4ede4d-3.0.0-py3-none-any.whl
                └── example_toml_b_4a4ede4d-3.0.0.tar.gz
        
        19 directories, 24 files
  
      ''',
    }),
    'stderr': '<not included>',
    'stdout': '''
      example-toml-4a4ede4d
  
    ''',
  })
# ---
# name: test_build_test_scenarios[example2]
  dict({
    'exit_code': 0,
    'filesystem': dict({
      'build/example-yaml-4692130b/example-yaml-4692130b-0.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-yaml-4692130b/example-yaml-4692130b-0.0.0/pyproject.toml': 'md5:47ff40e1cb004e6b8e2a196c83b3d5d5',
      'build/example-yaml-4692130b/example-yaml-4692130b-0.0.0/src/example_yaml_4692130b/__init__.py': '''
        __version__ = "0.0.0"
  
      ''',
      'build/example-yaml-4692130b/example-yaml-a-4692130b-1.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-yaml-4692130b/example-yaml-a-4692130b-1.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_yaml_a_4692130b"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_yaml_a_4692130b"]
        
        [project]
        name = "example-yaml-a-4692130b"
        version = "1.0.0"
        dependencies = [
            '''example-yaml-b-4692130b>1.0.0''',
        ]
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-yaml-4692130b/example-yaml-a-4692130b-1.0.0/src/example_yaml_a_4692130b/__init__.py': '''
        __version__ = "1.0.0"
  
      ''',
      'build/example-yaml-4692130b/example-yaml-b-4692130b-1.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-yaml-4692130b/example-yaml-b-4692130b-1.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_yaml_b_4692130b"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_yaml_b_4692130b"]
        
        [project]
        name = "example-yaml-b-4692130b"
        version = "1.0.0"
        dependencies = []
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-yaml-4692130b/example-yaml-b-4692130b-1.0.0/src/example_yaml_b_4692130b/__init__.py': '''
        __version__ = "1.0.0"
  
      ''',
      'build/example-yaml-4692130b/example-yaml-b-4692130b-2.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-yaml-4692130b/example-yaml-b-4692130b-2.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_yaml_b_4692130b"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_yaml_b_4692130b"]
        
        [project]
        name = "example-yaml-b-4692130b"
        version = "2.0.0"
        dependencies = []
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-yaml-4692130b/example-yaml-b-4692130b-2.0.0/src/example_yaml_b_4692130b/__init__.py': '''
        __version__ = "2.0.0"
  
      ''',
      'build/example-yaml-4692130b/example-yaml-b-4692130b-3.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-yaml-4692130b/example-yaml-b-4692130b-3.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_yaml_b_4692130b"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_yaml_b_4692130b"]
        
        [project]
        name = "example-yaml-b-4692130b"
        version = "3.0.0"
        dependencies = [
            '''example-yaml-c-4692130b''',
        ]
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-yaml-4692130b/example-yaml-b-4692130b-3.0.0/src/example_yaml_b_4692130b/__init__.py': '''
        __version__ = "3.0.0"
  
      ''',
      'dist/example-yaml-4692130b/example_yaml_4692130b-0.0.0.tar.gz': 'md5:dc8fef992b11413e901b42f77967e2f8',
      'dist/example-yaml-4692130b/example_yaml_a_4692130b-1.0.0-py3-none-any.whl': 'md5:93b4db1841920260969c8d624eeb021d',
      'dist/example-yaml-4692130b/example_yaml_a_4692130b-1.0.0.tar.gz': 'md5:f173d27690cf8ce11833c1928c05015f',
      'dist/example-yaml-4692130b/example_yaml_b_4692130b-1.0.0-py3-none-any.whl': 'md5:b71009b6c23a0e7835565faa1ea3b416',
      'dist/example-yaml-4692130b/example_yaml_b_4692130b-1.0.0.tar.gz': 'md5:9b9063658216582e8b6c28b4dcf12abd',
      'dist/example-yaml-4692130b/example_yaml_b_4692130b-2.0.0-py3-none-any.whl': 'md5:0acae420fd4d86d6b77355d436eb1e24',
      'dist/example-yaml-4692130b/example_yaml_b_4692130b-2.0.0.tar.gz': 'md5:2ae0e67e152c6b45af6dcedaeac5fc9f',
      'dist/example-yaml-4692130b/example_yaml_b_4692130b-3.0.0-py3-none-any.whl': 'md5:971df9a828a59dc0687634af3963faaf',
      'dist/example-yaml-4692130b/example_yaml_b_4692130b-3.0.0.tar.gz': 'md5:63cfe1d9f3276e8eb08606e44ec628e2',
      'tree': '''
        test_build_test_scenarios_exam2
        ├── build
        │   └── example-yaml-4692130b
        │       ├── example-yaml-4692130b-0.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── example_yaml_4692130b
        │       │           └── __init__.py
        │       ├── example-yaml-a-4692130b-1.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── example_yaml_a_4692130b
        │       │           └── __init__.py
        │       ├── example-yaml-b-4692130b-1.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── example_yaml_b_4692130b
        │       │           └── __init__.py
        │       ├── example-yaml-b-4692130b-2.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── example_yaml_b_4692130b
        │       │           └── __init__.py
        │       └── example-yaml-b-4692130b-3.0.0
        │           ├── .gitignore
        │           ├── pyproject.toml
        │           └── src
        │               └── example_yaml_b_4692130b
        │                   └── __init__.py
        └── dist
            └── example-yaml-4692130b
                ├── example_yaml_4692130b-0.0.0.tar.gz
                ├── example_yaml_a_4692130b-1.0.0-py3-none-any.whl
                ├── example_yaml_a_4692130b-1.0.0.tar.gz
                ├── example_yaml_b_4692130b-1.0.0-py3-none-any.whl
                ├── example_yaml_b_4692130b-1.0.0.tar.gz
                ├── example_yaml_b_4692130b-2.0.0-py3-none-any.whl
                ├── example_yaml_b_4692130b-2.0.0.tar.gz
                ├── example_yaml_b_4692130b-3.0.0-py3-none-any.whl
                └── example_yaml_b_4692130b-3.0.0.tar.gz
        
        19 directories, 24 files
  
      ''',
    }),
    'stderr': '<not included>',
    'stdout': '''
      example-yaml-4692130b
  
    ''',
  })
# ---
# name: test_view_test_scenarios[example0]
  dict({
    'exit_code': 0,
    'filesystem': dict({
      'tree': '''
        test_view_test_scenarios_examp0
        
        0 directories
  
      ''',
    }),
    'stderr': '',
    'stdout': '''
      example-961b4c22
      ├── environment
      │   └── python3.8
      ├── root
      │   └── requires a
      │       └── satisfied by a-1.0.0
      ├── a
      │   └── a-1.0.0
      │       └── requires b>1.0.0
      │           ├── satisfied by b-2.0.0
      │           └── satisfied by b-3.0.0
      └── b
          ├── b-1.0.0
          ├── b-2.0.0
          └── b-3.0.0
              └── requires c
                  └── unsatisfied: no versions for package
      
  
    ''',
  })
# ---
# name: test_view_test_scenarios[example1]
  dict({
    'exit_code': 0,
    'filesystem': dict({
      'tree': '''
        test_view_test_scenarios_examp1
        
        0 directories
  
      ''',
    }),
    'stderr': '',
    'stdout': '''
      example-toml-4a4ede4d
      ├── environment
      │   └── python3.8
      ├── root
      │   └── requires a
      │       └── satisfied by a-1.0.0
      ├── a
      │   └── a-1.0.0
      │       └── requires b>1.0.0
      │           ├── satisfied by b-2.0.0
      │           └── satisfied by b-3.0.0
      └── b
          ├── b-1.0.0
          ├── b-2.0.0
          └── b-3.0.0
              └── requires c
                  └── unsatisfied: no versions for package
      
  
    ''',
  })
# ---
# name: test_view_test_scenarios[example2]
  dict({
    'exit_code': 0,
    'filesystem': dict({
      'tree': '''
        test_view_test_scenarios_examp2
        
        0 directories
  
      ''',
    }),
    'stderr': '',
    'stdout': '''
      example-yaml-4692130b
      ├── environment
      │   └── python3.8
      ├── root
      │   └── requires a
      │       └── satisfied by a-1.0.0
      ├── a
      │   └── a-1.0.0
      │       └── requires b>1.0.0
      │           ├── satisfied by b-2.0.0
      │           └── satisfied by b-3.0.0
      └── b
          ├── b-1.0.0
          ├── b-2.0.0
          └── b-3.0.0
              └── requires c
                  └── unsatisfied: no versions for package
      
  
    ''',
  })
# ---
